lua/null-ls/helpers/command_resolver.lua

```
@@ -0,0 +1,56 @@
+local log = require("null-ls.logger")
+local s = require("null-ls.state")
+local u = require("null-ls.utils")
+
+local M = {}
+
+M.generic = function(params, prefix)
+    local resolved = s.get_resolved_command(params.bufnr, params.command)
+    -- a string means command was resolved on last run
+    -- false means the command already failed to resolve, so don't bother checking again
+    if resolved and (type(resolved.command) == "string" or resolved.command == false) then
+        log:debug(string.format("Using cached value for [%s] at %s", params.command, resolved.command))
+        return resolved.command
+    end
+
+    local executable_to_find = prefix and u.path.join(prefix, params.command) or params.command
+    log:debug("attempting to find local executable " .. executable_to_find)
+
+    local root = u.get_root()
+
+    local resolved_command, resolved_cwd
+
+    u.path.traverse_parents(params.bufname, function(dir)
+        resolved_command = u.path.join(dir, executable_to_find)
+        if u.is_executable(resolved_command) then
+            log:trace(string.format("resolved dynamic command for [%s] with cwd=%s", executable_to_find, dir))
+            resolved_cwd = dir
+            return true
+        end
+
+        resolved_command = nil
+        resolved_cwd = nil
+        -- use cwd as a stopping point to avoid scanning the entire file system
+        if dir == root then
+            return true
+        end
+    end)
+
+    if resolved_command then
+        local is_executable, err_msg = u.is_executable(resolved_command)
+        if not is_executable then
+            log:warn(err_msg)
```

We should throw an error here or handle this case to prevent caching a `resolved_command` that isn't executable.